pkgs/servers/http/apache-modules/mod_ca/default.nix

```diff
@@ -0,0 +1,33 @@
+{ stdenv, fetchurl, gnused, coreutils, pkgconfig, apacheHttpd, openssl, openldap }:
+
+stdenv.mkDerivation rec {
+ baseurl = "https://redwax.eu/dist/rs/";
+ name = "mod_ca";
+ suffix = ".tar.gz";
+
+ meta = with stdenv.lib; {
+   description = "RedWax CA service modules.";
+   version = "0.2.1";
+
+   homepage = "https://redwax.eu";
+   license = licenses.asl20;
+   platforms = platforms.unix;
+   maintainers = with maintainers; [ dirkx ];
+ };
+
+ src = fetchurl {
+   url = "${baseurl}${name}-${meta.version}${suffix}";
+   sha256 = "1pxapjrzdsk2s25vhgvf56fkakdqcbn9hjncwmqh0asl1pa25iic";
+ };
+
+ buildInputs = [ gnused coreutils pkgconfig apacheHttpd openssl openldap ];
+
+ configurePlatforms = [];
+ configureFlags = [
+       "--with-apxs=${apacheHttpd.dev}/bin/apxs"
+	];
+
+ installPhase = "make INCLUDEDIR=$out/include LIBEXECDIR=$out/libexec install";
```

Comment:
Maybe we could put this in `installFlags` like so:
```
  installFlags = [
    "INCLUDEDIR=$(out)/include"
    "LIBEXECDIR=$(out)/modules"
  ];
```

Note that by convention all `mod_*.so` files should go under `${out}/modules`.

Reply:
So we've done it in this particular way - as it then cleanly overlays with the Apache-httpd (these are modules for apache) and Apache its default path finds these there. So including it is pretty much no configuration file/change/etc - can even be a cmd line flag from systemd.

Moving to modules means that you can no longer rely on that path - and need to be extra careful when sandboxing.

Or is the pattern in https://github.com/NixOS/nixpkgs/blob/465566948393cf533e3617704d1c4ccc34cf3753/pkgs/desktops/gnome-3/core/gnome-user-share/default.nix - line 46 the correct one -- and should any users of this module not use the default configs - but tweak all the paths explicitly for all their modules ?

And should I change this to modules as you suggest (and also do a pull request for apache ?)

Reply:
If I take a look at some standard config shipped with `Debian they use `modules/`:
```
cat /etc/apache2/mods-enabled/php7.3.load
# Conflicts: php5
# Depends: mpm_prefork
LoadModule php7_module /usr/lib/apache2/modules/libphp7.3.so
```

I'm pretty sure I've seen `libexecdir/` or some such too... maybe on `RHEL`?

Also `apacheHttpd` via `services.httpd` on NixOS won't automatically find any of your apache modules which aren't part of the `apacheHttpd` package itself. You need to explicitly tell the `httpd` service about them via the `services.httpd.extraModules` option like so:
```
services.httpd.extraModules = [
  { name = "mod_ca"; path = "${pkgs.apacheHttpdPackages}/modules/mod_ca.so"; }
];
```

Notice that I have to specify the whole path. So you can use either `libexecdir` or `modules`, but most packages use `modules` in nixpkgs. No hard requirement.

What would your PR against apache look like? Are you talking about the package or the service?

PS. Actually typing this out makes me realize how tedious the whole `.extraModules path` thing is... PR incoming are some point to make that easier and less error prone.

Reply:
Well - it gets messier even - as you also have the include part - and the need for APXS to build the libraries in a way that apache can dlopen load them (i.e. get flags like -fPIC not from CFLAGS but from the CFLAGS and other flags used at the time of the apache compile). 

So APXS, as it comes with apache, has its own hardcode include and libexec locations (it predates pkgconfig)  -- and the includes are needed across package boundaries (in this case - we need the apache include files and the mod_ca files. 

Se we use a bit of a trick here to bypass APXS by a modified Makefile.am. But do not try to second guess APXS (by using normal gcc/clang compiles with identical flags). So as thin as possible.

Reply:
Is this something specific to the apache modules you're packaging, or you mean in general? I haven't ran into any issues with any of the apache modules in nixpkgs that I have used. :thinking:

Reply:
So we do something a bit unusual and that is we have a main apache module (mod_ca) and then modules that need both the apache dl/apxs -and- mod-ca. So that means that you are not just runtime pulling dl's from any path - but also includes during build. So this is a bit like why you somethings see a libperl or libpython next to a mod_perl stub.

Regardless - we cannot install them in the APXS provided location for modules -- as this one is private to httpd (/nix/store/m15l5xmr2s02sji9yh91nip6b6hrr8r4-apache-httpd-2.4.41/modules/mod_include.so).

But no problem to change this to modules - and then do the same 'sed' substution as for exampled used in [other places](https://github.com/NixOS/nixpkgs/blob/465566948393cf533e3617704d1c4ccc34cf3753/pkgs/desktops/gnome-3/core/gnome-user-share/default.nix#L45) to adapt it at the using side.

Reply:
I have never used `gnome-user-share` but after quickly looking through the derivation and upstream documentation I'm left with the impression that `gnome-user-share` possibly spawns an `apacheHttpd` executable and provides configuration to instruct `apacheHttpd` where apache modules are. This is very separate from how the `services.httpd` nixos module works.

Maybe @jtojnar or @worldofpeace might have some more insight into this, as the resident `gnome` experts, and all around very bright :bulb: people.

Reply:
I've moved it to modules - and tested that I can tweak both cyrus/imapd and wordpress sufficently to be happy. So no issues. Will be fine & can re-visit once we have done the other modules too.

Reply:
I think there is a value in mirroring the FHS structure inside `$out` so that people coming from other distros can quickly understand the output tree. We have been already doing such transitions (https://github.com/NixOS/nixpkgs/pull/43992, https://github.com/NixOS/nixpkgs/pull/30428). Of course it is harder when no such standard location exists. On Arch [`apache`](https://www.archlinux.org/packages/extra/x86_64/apache/) uses `lib/httpd/modules`, whereas [`lighttpd`](https://www.archlinux.org/packages/extra/x86_64/lighttpd/) uses `lib/lighttpd`. It definitely should not be in `$out/modules`, though.

Indeed `gnome-user-share` runs its own `httpd` instance but I would not say it is that different from NixOS module, it is just another instance of extensibility conflicting with immutable store, a well understood issue. The only difference is that in the module, we are creating a configuration from scratch so we can just put in whatever module paths we want, whereas in `gnome-user-share`, we are re-using the upstream config so we need to patch the paths there to combat the assumption that all modules are in a single directory.

Reply:
And on Debian it is `lib/apache2/modules`. I suppose Debian wasn't really a great example for me to base this on, given the extremes Debian goes to modify `httpd` from upstream. I would guess `lib/httpd/modules` is probably the most standard path and we should go with that.

@jtojnar maybe I should make a PR to utilize `passthru` and add something like `moduleName` and `modulePath` so the `services.httpd.extraModules` can accept a package and always know what to do, similar to what the `odbc` drivers do. This would be a nice addition to the standardization you propose, all in one PR. What do you think?

Reply:
Sounds good to me.